### PR TITLE
Add visualizers for padding and margin for all blocks

### LINKED ISSUE #36418: Full Site Editing: Hovering block padding and margin inputs are registered as changes
### Description

Putting the mouse cursor on top of some block attributes inputs (e.g. padding, margin) enables the "Save" button on the top right, even though nothing changed.

This behavior exists because of the cover block padding and margin visualizer, introduced on #23041.

Note that we do want to keep the visualizer showing up when hovering over the input. What we don't want is to enable the "Save" button when this happens.

A possible solution would be to introduce a new piece of state that would host the visualizer settings itself, not polluting the changes object and not triggering the false alarm.

### Step-by-step reproduction instructions

1. Enable the full site editor locally;
2. Navigate to full site editor;
3. Select any block that has spatial settings and hover the padding or margin inputs;
4. Check that the "Save" button is enabled.

### Screenshots, screen recording, code snippet

https://user-images.githubusercontent.com/26530524/141362112-f68d3e30-8134-4682-a6e3-be87079f7fe0.mov

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### LINKED ISSUE #40057: BoxControlVisualizer: Create a new separate component, and fix existing issues
## What problem does this address?

We'd like to add more visual feedback on the canvas when using Gutenberg design tools. Here's an example of how the existing BoxControlVisualizer can be used to display padding on an element on the cavas:

Screen Shot 2022-04-05 at 1 52 10 PM

Currently the [BoxControlVisualizer](https://github.com/WordPress/gutenberg/blob/028bdeeba6e4b37b02b13b3fa543982cb579f71e/packages/components/src/box-control/visualizer.js) is in an unusable state and is turned off. It's also tied closely with the BoxControl component. Some of the recently highlighted issues include:

* An issue caused by block attributes storing visualizer data: https://github.com/WordPress/gutenberg/issues/31839
* An attempt to switch the visualizer to using state: https://github.com/WordPress/gutenberg/pull/33560
* A comment from Riad regarding visualizers, switching them using popovers etc: https://github.com/WordPress/gutenberg/issues/36418#issuecomment-972570449

## What is your proposed solution?

I think we should:

- [ ] Extract the BoxControlVisualizer into its own component so we can convert to TypeScript, add tests, and develop and review it in isolation.
- [ ] Modify the visualizer so that it uses popover or :before :after styles so that we aren't adding a bunch of wrapper markup to the DOM to support it.
- [ ] Look at using state instead of block attributes to support the new component.
- [ ] Test this new visualizer component with a different design tool, perhaps block gap.

### PR DESCRIPTION
**Pull request by youknowriad**
closes #36418 
closes #40057

## What?

This PR adds "visualizers" when editing padding/margin controls in blocks that support them.

## Why?

The visualizers gives the user clarity of the impact its changes are having on the block.

## How?

Previously we had `BoxControl.Visualizer` components, it was used in the Cover block padding for instance, but it suffered two major issues:

 - It relied on rendering a "div" in the DOM inside the block itself. The problem is that this impacts the layout and style of the blocks (causes layout shifts...), it's not a technique we can use in all blocks. The current PR solves this by making the visualizer work like the block toolbar using a "Popover". It uses a special Popover that tries to mirror the size of the block to cover it.
 - The second issue is that the visualizers relied on block "attribute" changes to support showing them on "hover". This causes issues like creating "undo" levels, making the "post" dirty and keeping a trace in the saved markup as soon as you "hover" a padding control. I elected to remove that entirely and only show the visualizers on "edit" right now. If it's deemed not sufficient, I can look at restoring the hover effect without using the attributes.

**Note**

I'm planning to use the BlockPopover (with cover argument) in #40376 to render the side remove buttons... So this PR is also a requirement for that.

## Testing Instructions

1- Insert a block that supports padding or margin (group, cover...)
2- Tweak the value for these controls
3- Notice the visual effect (blue squares) when changing the value.

## Screenshots or screencast 

https://user-images.githubusercontent.com/272444/164422610-e01d5904-4f52-4aaf-93bd-ea4abdd5a70c.mov

### CODE REVIEW

**Comment by youknowriad** [comment]
@aaronrobertshaw I noticed that you've added a new visualizer for border, maybe we can use the same technique as this PR? How can I experience it?

**Comment by Mamaduka** [comment]
Does this supersede #40253, or will the new `BoxModelOverlay` be used here when it's ready?

**Comment by aaronrobertshaw** [comment]
> I noticed that you've added a new visualizer for border, maybe we can use the same technique as this PR? How can I experience it?

Thanks for the ping. @kevin940726 is currently working on https://github.com/WordPress/gutenberg/pull/40253 which reimagines the visualizer. I believe the plan was to support borders via that.

I'm not sure we can call the new [`BorderControl`](https://github.com/WordPress/gutenberg/pull/37769) and [`BorderBoxControl`](https://github.com/WordPress/gutenberg/pull/38876) components visualizers for borders. I'm open to any ideas on improving visual feedback in the editor though.

**Comment by youknowriad** [comment]
I actually didn't know about https://github.com/WordPress/gutenberg/pull/40253 :) 

In principle, it's the same thing but I think "Popover" is already capable of handing what the new component is trying to do without the added code and complexity. (I do think there are opportunities to improve the internals of Popover) though.

**Comment by jasmussen** [comment]
This is just delightful. The visualizers of paddings and margins both work great, but notably for margins it might finally unblock the addition of margin tools to more blocks since it finally properly visualizes how they work, and surface any margin-collapsing that might otherwise make margins not take. All shown in this GIF:
visualizers

This is all working pretty well as-is, though I encountered one issue. In this GIF I group two paragraphs, apply a bunch of padding, and while focus is still inside the padding input field, I press ⌘Z to undo, and it doesn't undo the padding change and it makes the padding flash:

undo

Something similar but not quite the same happens when I press the explicit undo button in the toolbar.

**Comment by youknowriad** [comment]
@jasmussen I haven't been able to reproduce neither of these issues 🤔

**Comment by jasmussen** [comment]
I can still reproduce after a fresh npm install. Try these steps exactly:

* Select a group with contents
* Click and drag upwards in the Padding control to increase the padding
* Press ⌘Z on the keyboard multiple times

What do you see?

**Comment by youknowriad** [comment]
@jasmussen I reproduced once, but I'm failing to reproduce now 😬

**Comment by youknowriad** [comment]
@jasmussen oh, I see now, It is actually a bug in trunk as well, it's independent of this PR. Basically when I try to undo a change of padding, it undo but then restores the old value right away. I think it's probably a bug in `BoxControl`. I'll try to debug it and fix it separately.

**Comment by youknowriad** [comment]
@jasmussen the fix for the padding change is here https://github.com/WordPress/gutenberg/pull/40518 but I might need help from components folks to make sure the tests and everything is correct.

I think we shouldn't block the work in the current PR though since the bug is already in trunk.

**Comment by jasmussen** [comment]
Definitely agree, this one feels amazing. I'd be happy to give a green check.

**Comment by mtias** [comment]
I love this is being accomplished with more code removed than added :)

File: packages/block-editor/src/hooks/margin.js [hunk 1]
```diff
@@ -155,3 +149,58 @@ export function MarginEdit( props ) {
...
+	const style = useMemo( () => {
+		return {
+			borderTopWidth: margin?.top ?? 0,
+			borderRightWidth: margin?.right ?? 0,
+			borderBottomWidth: margin?.bottom ?? 0,
+			borderLeftWidth: margin?.left ?? 0,
+			top: margin?.top ? `-${ margin?.top }` : 0,
+			right: margin?.right ? `-${ margin?.right }` : 0,
+			bottom: margin?.bottom ? `-${ margin?.bottom }` : 0,
+			left: margin?.left ? `-${ margin?.left }` : 0,
```
Comment (mcsf) [inline packages/block-editor/src/hooks/margin.js]:
```suggestion
			top: margin?.top ? `-${ margin.top }` : 0,
			right: margin?.right ? `-${ margin.right }` : 0,
			bottom: margin?.bottom ? `-${ margin.bottom }` : 0,
			left: margin?.left ? `-${ margin.left }` : 0,
```

File: packages/block-library/src/cover/edit.js [hunk 2]
```diff
@@ -744,11 +740,6 @@ function CoverEdit( {
 				style={ { ...style, ...blockProps.style } }
 				data-url={ url }
 			>
-				<BoxControlVisualizer
-					values={ styleAttribute?.spacing?.padding }
-					showValues={ styleAttribute?.visualizers?.padding }
-					className="block-library-cover__padding-visualizer"
-				/>
```
Comment (mcsf) [inline packages/block-library/src/cover/edit.js:751]: 👏

**Review by mcsf — APPROVED** [review]
I like the approach, and this seems to be working well.

As far as implementation details go, I like the locality: the relationship between the hooks, BlockPopover and Popover is fairly intuitive, and it's clear what each piece is responsible for.

Bonus points for cutting down on unused components and converging. :)

**Comment by youknowriad** [comment]
Going to merge this as it's gets the feature in without new APIs or overhead but it doesn't mean we can't extract reusable components from this later if needed.

**Pull request merged**